File: src/y2024/Day22.kt
package y2024

import kotlin.time.measureTimedValue

fun main() {

    fun generatePrices(seed: Long) =
        generateSequence(seed) {
            var num = (it shl 6 xor it) % 16777216
            num = (num shr 5 xor num) % 16777216
            num = (num shl 11 xor num) % 16777216
            num
        }.take(2001)

    fun part1(input: List<String>) = input.sumOf { generatePrices(it.toLong()).last() }

    fun part2(input: List<String>) = input.asSequence()
        .flatMap {
            generatePrices(it.toLong())
                .map { it.toInt() % 10 }
                .zipWithNext { a, b -> b - a to b }
                .windowed(4) { it.map { it.first } to it.last().second }
                .distinctBy { it.first }
        }
        .groupBy({ it.first }, { it.second })
        .maxOf { it.value.sum() }

    val testInput = readInput("Day22_test")
    val testInput2 = readInput("Day22_test2")
    check(part1(testInput).apply { println() } == 37327623L)
    check(part2(testInput2).apply { println() } == 23)

    val input = readInput("Day22")
    measureTimedValue { part1(input) }.println()
    measureTimedValue { part2(input) }.println()
}

File: src/y2024/Day22.py
from itertools import islice
from itertools import pairwise
from itertools import groupby
from time import time


def window(seq, n=4):
    it = iter(seq)
    result = tuple(islice(it, n))
    if len(result) == n:
        yield result
    for elem in it:
        result = result[1:] + (elem,)
        yield result


def generate_prices(seed: int):
    yield seed
    for _ in range(2000):
        seed = (seed << 6 ^ seed) % 16777216
        seed = (seed >> 5 ^ seed) % 16777216
        seed = (seed << 11 ^ seed) % 16777216
        yield seed


def part1(input_list):
    return sum(next(islice(generate_prices(int(x)), 2000, 2001)) for x in input_list)


def part2(input_list):
    diffs = []

    for x in input_list:
        sequence = [n % 10 for n in generate_prices(int(x))]
        pairs = [(b - a, b) for a, b in pairwise(sequence)]
        windowed = [((a[0], b[0], c[0], d[0]), d[1]) for a, b, c, d in window(pairs)]

        unique = {}
        for first, second in windowed:
            if first not in unique:
                unique[first] = second

        diffs.append(unique)

    grouped = {}
    for unique in diffs:
        for key, value in unique.items():
            grouped.setdefault(key, []).append(value)

    return max(sum(values) for values in grouped.values())


def read_input(filename):
    with open(filename) as f:
        return [line.strip() for line in f]

test_input = read_input("Day22_test.txt")
test_input2 = read_input("Day22_test2.txt")
p1 = part1(test_input)
print(p1)
assert p1 == 37327623
p2 = part2(test_input2)
print(p2)
assert p2 == 23

input_data = read_input("Day22.txt")
start = time()
print(part1(input_data), f"| Execution time: {time() - start:.6f}s")

start = time()
print(part2(input_data), f"| Execution time: {time() - start:.6f}s")
